AGDNS-4106-mv-articles-to-dns-kb-vol.2 (#443)
AdguardTeam/KnowledgeBaseDNS: docs/adguard-home/dhcp.md

---
title: DHCP
sidebar_position: 10
---

AdGuard Home can be used as a DHCP server. This page describes how to do that.

## Prerequisites {#prerequisites}

> 1. Make sure that you run an OS on which AdGuard Home supports DHCP. We currently don't support DHCP on Windows.

Reviewer: Here and below, Unicode apostrophe.

2. Make sure that your machine has a static IP address.

> ## Default options {#default-options}

Reviewer: This doesn't seem like something that needs its own second-level section.  Perhaps, merge into the next one as a simple paragraph?

By default, AdGuard Home will set itself as the DNS server for the DHCP clients. The default lease time is 24 hours.

## Configuration {#configuration}

See the DHCP section in the [configuration] article for the overview of the DHCP configuration options. There are several configuration parameters for DHCP that can't be set via the AdGuard Home administrator dashboard. Those are described below.

> ### DHCPv4 options {#dhcpv4-options}

Reviewer 2: I think it would be reasonable to reference this exact header instead of the whole page in the `configuration.md` page.  The same applies to the `{#dhcpv6-options}`.

### DHCPv6 options {#dhcpv6-options}

The option `dhcp.dhcpv6.ra_slaac_only`, if `true`, sends RA packets forcing the clients to use SLAAC. The DHCPv6 server won't be started in this case.

The option `dhcp.dhcpv6.ra_allow_slaac`, if `true`, sends RA packets allowing the clients to choose between SLAAC and DHCPv6.

[configuration]:    https://adguard-dns.io/kb/adguard-home/configuration/
[rfc-1191-tbl-7.1]: https://datatracker.ietf.org/doc/html/rfc1191#section-7.1
[rfc-2131]:         https://datatracker.ietf.org/doc/html/rfc2131
[rfc-2132]:         https://datatracker.ietf.org/doc/html/rfc2132

## Automatic hosts {#auto-hosts}

Machines in the network can be reached more easily using the hostnames they send in the DHCP requests with a configurable top-level domain (TLD). By default, the TLD is `lan`. For example, if you have a machine called “workstation” in the network, and it sends a DHCP request with option 12 set to `workstation`, you can reach it over HTTP on the host `http://workstation.lan`.

> You can also set a custom TLD or domain name using the `dns.local_domain_name` field in the [configuration] file.

Reviewer 2: We've moved it to the `dhcp` section, AFAIK.  The reference could be more specific.

---
title: DHCP
sidebar_position: 10
---

AdGuard Home can be used as a DHCP server. This page describes how to do that.

## Prerequisites {#prerequisites}

1. Make sure that you run an OS on which AdGuard Home supports DHCP. We currently don’t support DHCP on Windows.

2. Make sure that your machine has a static IP address.

## Configuration {#configuration}

> See the DHCP section in the [configuration][dhcp-conf] article for the overview of the DHCP configuration options. There are several configuration parameters for DHCP that can’t be set via the AdGuard Home administrator dashboard. Those are described below.

Reviewer 3: ```suggestion
See the [overview of the DHCP configuration options][dhcp-conf]. There are several parameters for DHCP that can’t be set via the AdGuard Home administrator dashboard. Those are described below.
```

:::note

By default, AdGuard Home will set itself as the DNS server for the DHCP clients. The default lease time is 24 hours.

:::

[dhcp-conf]: /adguard-home/configuration#dhcp

### DHCPv4 options {#dhcpv4-options}

The `options` field is used to explicitly specify the values for DHCP options and modify the response. In accordance with *Section 4.3.1* of [RFC 2131][rfc-2131], these options override the default options’ values set by Adguard Home and requested by a client, which means that if you want to set custom DNS server addresses using option `6` (Domain Name Server), you may want also add Adguard Home’s own addresses there. Otherwise, AdGuard Home’s filtering won’t work for the DHCP clients who receive these DNS server addresses.

> Any option begins with an option *code* written as decimal integer. See [RFC 2132][rfc-2132] for the actual DHCP option codes and allowed lengths. The code is followed by an option’s *type* and *value*. Currently the following *types* are supported:

Reviewer 3: ```suggestion
Any option begins with a code written as a decimal integer. See [RFC 2132][rfc-2132] for the actual DHCP option codes and allowed lengths. The code is followed by the option’s type and value. Currently the following types are supported:
```

- `bool`: Human-readable form of a boolean value, and has the length of 1 octet.

    **Example**:

    ```yaml
    'options':
    - '19 bool 0'     # Disable IP forwarding for hosts.
    - '20 bool t'     # Enable non-local source routing for hosts.
    - '29 bool F'     # Disable subnet mask discovery.
    - '30 bool true'  # Enable mask supplying for supporting hosts.
    - '36 bool False' # Make the hosts use RFC 894 for ethernet encapsulation.
    ```

> - `del`: No-value option and is used to unconditionally remove options from the server’s responses (which may lead to weird behaviors, use with caution).

Reviewer 3: it should be "No-value option is used", I guess

    Since the list of options is interpreted sequentially from first to last, the subsequent option may override the previous ones. So this:

    ```yaml
    'options':
    - '19 bool T'
    - '19 del'
    - '20 del'
    - '20 bool F'
    ```

    instructs to remove the option `19`, and to set the option `20` to `false`.

> - `dur`: Human-readable form of a duration in range [0 – 4294967296 seconds (about 136 days)] and has a length of *4* octets, just like a 32-bit unsigned integer.

Reviewer 3: ```suggestion
- `dur`: A human-readable form of a duration in the range of 0 to 4,294,967,296 seconds (approximately 136 days), consisting of *4* octets, just like a 32-bit unsigned integer.
```

    **Example**:

    ```yaml
    'options':
    - '24 dur 10m'
    ```

> - `hex`: Sequence of hexadecimal numbers of an arbitrary length.

Reviewer 3: - `hex`: A sequence of hexadecimal numbers of arbitrary length.

    **Example**:

    ```yaml
    'options':
    - '25 hex 0044012801FC03EE05D407D211001FE645FA'
    ```

- `ip`: Accepts an IPv4 address and has a length of *4* octets, just like an IPv4 itself.

    **Example**:

    ```yaml
    'options':
    - '28 ip 192.168.0.255'
    ```

- `ips`: Accepts a comma-separated list of IPv4 addresses. It has an arbitrary length, but is always a multiple of *4* octets.

    **Example**:

    ```yaml
    'options':
    - '6 ips 1.2.3.4,1.2.3.5'
    ```

- `text`: Accepts an arbitrary UTF-8 encoded string and has a length of encoded text.

    **Example:**

    ```yaml
    'options':
    - '252 text http://server.domain/proxyconfig.pac'
    ```

> - `u8`: Decimal number in range [0 – 255] and takes *1* octet, just like an unsigned 8-bit integer.

Reviewer 3: A decimal number in the range of 0 to 255 that takes *1* octet, just like an unsigned 8-bit integer.

    **Example:**

    ```yaml
    'options':
    - '23 u8 64'
    ```

> - `u16`: Decimal number in range [0 – 65535] and takes *2* octets, just like an unsigned 16-bit integer.

Reviewer 3: same

| All subnets are local            | False                                         |
| Perform Mask Discovery           | False                                         |
| Mask Supplier                    | False                                         |
| Perform Router Discovery         | True                                          |
| Router Solicitation Address      | 224.0.0.2                                     |
| Broadcast Address                | 255.255.255.255                               |
| Use Trailer Encapsulation        | False                                         |
| ARP Cache Timeout                | 1 minute                                      |
| Ethernet Encapsulation version   | RFC 894                                       |
| Default TCP TTL                  | 60 seconds                                    |
| TCP Keepalive Interval           | 2 hours                                       |
| Put TCP Keepalive Garbage        | True                                          |
| Routers                          | `gateway_ip` from configuration               |
| Subnet Mask                      | `subnet_mask` from configuration              |

> Some of these values may appear obsolete or may cause issues with some DHCP client implementations among the many existing. In accordance with [RFC 2131][rfc-2131] the options, when not explicitly configured, are only returned if requested by client within the option `55` (Parameter Request List).

Reviewer 3: Some of these values may appear obsolete or cause issues with certain DHCP client implementations. According to [RFC 2131][rfc-2131], the options are only returned if requested by the client within option `55` (Parameter Request List) when not explicitly configured.
